naga/src/back/continue_forward.rs

```diff
@@ -0,0 +1,204 @@
+//! Workarounds for platform bugs in switches and loops.
+//!
+//! In these docs, we use CamelCase links for Naga IR concepts, and ordinary
+//! `code` formatting for HLSL or GLSL concepts.
+//!
+//! ## Avoiding `continue` within `switch`
+//!
+//! As described in <https://github.com/gfx-rs/wgpu/issues/4485>, the FXC HLSL
+//! compiler doesn't allow `continue` statements within `switch` statements, but
+//! Naga IR does. We work around this by introducing synthetic boolean local
+//! variables and branches.
+//!
+//! Specifically:
+//!
+//! - We generate code for [`Continue`] statements that sets a `bool` local to
+//!   `true` and does a `break`.
+//!
+//! - When a [`Switch`] statement contains such a [`Continue`] statement:
+//!
+//!   - If it's the outermost such [`Switch`] within the targeted [`Loop`], we
+//!     declare the `bool` local, initialized to `false`, and immediately after
+//!     the `switch`, we check the local and do a `continue` if it's set.
+//!
+//!   - If the [`Switch`] is nested within other [`Switch`]es, then after the
+//!     generated `switch`, we check the local (which we know was declared
+//!     before the surrounding `switch`) and do a `break` if it's set.
```

Comment:
As a nit, I would say that this line might be clarified:

> When a [`Switch`] statement contains such a [`Continue`] statement:

To me, this could imply that the following points are only applied for switches containing continue statements, when it is really applied for all switches (that are inside a loop and can't use `continue` directly) since we don't do anything to detect whether a switch statement contains `continue`s.

But FWIW I don't think the false implication is very strong or significant to understanding the mechanisms here and I'm happy with the current text.

Reply:
It's a fair point - it might make people expect we do such a scan somewhere. I made the docs describe what we really do.